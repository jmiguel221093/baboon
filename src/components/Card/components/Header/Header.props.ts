import type { ReactNode } from "react";
import type { BasicIconActionType, IconType } from "../../../../types";

import type { MenuListItemProps } from "../../../MenuList";

type ActionsGroupType = {
	/** The icon of the button that groups the actions */
	icon?: ReactNode | IconType;
	/** The actions that will be grouped */
	actions?: MenuListItemProps[];
};

type CardMediaImageType = {
	source: string;
	alt?: string;
};

export interface HeaderProps {
	title?: string;
	metadata?: ReactNode;
	actionsGroup?: ActionsGroupType;
	mainActions?: BasicIconActionType[];
	mediaImage?: CardMediaImageType;
	headerInMedia?: boolean;
	icon?: ReactNode | IconType;
}
