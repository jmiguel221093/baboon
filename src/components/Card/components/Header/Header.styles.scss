$card-header-title-wrapper-margin: var(--spacing-1);
$card-header-media-image-wrapper-negative-spacing: var(--spacing-6);
$card-header-media-image-wrapper-spacing: var(--spacing-8);
$card-header-media-image-wrapper-border-radius: var(--border-radius-12);
$card-header-media-image-wrapper-header-color: var(--color-text-on-dark);
$card-header-media-image-wrapper-action-shadow-focus: 0 0 0 4px
	rgba(255, 255, 255, 0.5);

.Card__Header {
	&__TitleActionsWrapper {
		display: flex;
		align-items: center;
		flex: 1 1 0;
		.Card__Header {
			&__TitleWrapper {
				flex: 1 1 0;
			}
			&__MetadataWrapper {
				&:not(:first-child) {
					margin-top: $card-header-title-wrapper-margin;
				}
			}
		}
	}
	&__MediaImageWrapper {
		position: relative;
		border-radius: $card-header-media-image-wrapper-border-radius;
		margin-top: calc(
			$card-header-media-image-wrapper-negative-spacing * -1
		);
		margin-left: calc(
			$card-header-media-image-wrapper-negative-spacing * -1
		);
		margin-right: calc(
			$card-header-media-image-wrapper-negative-spacing * -1
		);
		overflow: hidden;
		&:not(:last-child) {
			margin-bottom: $card-header-media-image-wrapper-spacing;
		}
		&--has-content {
			&:before {
				content: "";
				display: block;
				width: 100%;
				height: 40%;
				position: absolute;
				left: 0;
				bottom: 0;
				background: rgb(2, 0, 36);
				background: linear-gradient(
					0deg,
					rgba(2, 0, 36, 0.5) 0%,
					rgba(0, 0, 0, 0) 100%
				);
				z-index: 0;
			}
		}
		img {
			display: block;
			max-width: 100%;
		}
		.Card__Header {
			&__TitleActionsWrapper {
				position: absolute;
				width: 100%;
				bottom: 0;
				left: 0;
				z-index: 1;
				padding: $card-header-media-image-wrapper-negative-spacing;
				button {
					color: $card-header-media-image-wrapper-header-color;
					&:hover {
						background-color: rgba(#fff, 0.3);
					}
					&:focus {
						background-color: rgba(#fff, 0.6);
						box-shadow: $card-header-media-image-wrapper-action-shadow-focus;
					}
				}
			}
			&__TitleWrapper {
				color: $card-header-media-image-wrapper-header-color;
			}
		}
	}
}
