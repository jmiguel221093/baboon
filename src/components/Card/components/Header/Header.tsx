import { useCallback, useState } from "react";
import { cssClassNames, isValidIcon } from "../../../../utils";
import { Button } from "../../../Button";
import { Icon } from "../../../Icon";
import { MenuList } from "../../../MenuList";
import { Popover } from "../../../Popover";
import { Stack } from "../../../Stack";
import { Text } from "../../../Text";
import type { HeaderProps } from "./Header.props";

import "./Header.styles.scss";

const COMPONENT_NAME = "Card__Header";

const Header = ({
	title,
	metadata,
	mainActions = [],
	actionsGroup,
	mediaImage,
	headerInMedia,
	icon,
}: HeaderProps) => {
	const [isActionsGroupOpen, setIsActionsGroupOpen] = useState(false);
	const isHeaderInMedia = mediaImage && headerInMedia;

	/* eslint-disable no-extra-parens */
	const iconSource = isValidIcon(icon) ? (
		<Icon source={icon} backdrop color="neutral" size="extra-large" />
	) : (
		icon
	);
	/* eslint-enable no-extra-parens */

	const className = cssClassNames(
		COMPONENT_NAME,
		isHeaderInMedia && "header-in-media"
	);

	const triggerActionsGroupPopover = useCallback(() => {
		setIsActionsGroupOpen((open) => !open);
	}, [actionsGroup]);

	const actionsGroupTrigger = actionsGroup ? (
		<Button
			icon={actionsGroup.icon}
			clean
			size="small"
			onClick={triggerActionsGroupPopover}
		/>
	) : null;

	const actionsGroupMarkup =
		actionsGroup && actionsGroupTrigger ? (
			<Popover
				open={isActionsGroupOpen}
				trigger={actionsGroupTrigger}
				onClose={triggerActionsGroupPopover}
				align="right"
			>
				<MenuList items={actionsGroup.actions} />
			</Popover>
		) : null;

	const actionsMarkup =
		actionsGroupMarkup || mainActions.length ? (
			<Stack spacing="medium">
				{actionsGroupMarkup}
				{mainActions.map((action, index) => (
					<Button
						key={action.id || index}
						{...action}
						clean
						size="small"
					/>
				))}
			</Stack>
		) : null;

	const iconMarkup = icon ? (
		<div className={`${COMPONENT_NAME}__IconWrapper`}>
			{iconSource as React.ReactNode}
		</div>
	) : null;

	const titleMarkup =
		title || actionsMarkup ? (
			<div className={`${COMPONENT_NAME}__TitleWrapper`}>
				<Text variant="heading">{title}</Text>
				{actionsMarkup}
			</div>
		) : null;

	const metadataMarkup = metadata ? (
		<div className={`${COMPONENT_NAME}__MetadataWrapper`}>{metadata}</div>
	) : null;

	const titleActionsMarkup =
		titleMarkup || metadata ? (
			<div className={`${COMPONENT_NAME}__TitleActionsWrapper`}>
				{iconMarkup}
				<div className={`${COMPONENT_NAME}__HeaderMainInfoWrapper`}>
					{titleMarkup}
					{metadataMarkup}
				</div>
			</div>
		) : null;

	const mediaImageMarkup = mediaImage ? (
		<div
			className={`${COMPONENT_NAME}__MediaImageWrapper${
				isHeaderInMedia
					? ` ${COMPONENT_NAME}__MediaImageWrapper--has-content`
					: ""
			}`}
		>
			<img src={mediaImage.source} alt={mediaImage.alt || title} />
			{isHeaderInMedia && titleActionsMarkup}
		</div>
	) : null;

	return (
		<div className={className}>
			{mediaImageMarkup}
			{!isHeaderInMedia && titleActionsMarkup}
		</div>
	);
};

export default Header;
