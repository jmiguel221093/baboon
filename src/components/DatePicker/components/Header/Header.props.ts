import type { IconProp } from "../../../../types";

export interface DatePickerHeaderProps {
	month: number;
	year: number;
	onPreviousClick?: () => void;
	onNextClick?: () => void;
	previousIcon?: IconProp;
	nextIcon?: IconProp;
}
