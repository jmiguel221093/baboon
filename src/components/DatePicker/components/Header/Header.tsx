import { cssClassNames } from "../../../../utils";
import { Button } from "../../../Button";
import { Text } from "../../../Text";
import { MONTHS } from "../../utils";
import type { DatePickerHeaderProps } from "./Header.props";

const COMPONENT_NAME = "DatePicker__Header";

const Header = ({
	month,
	year,
	onPreviousClick,
	onNextClick,
	previousIcon,
	nextIcon,
}: DatePickerHeaderProps) => {
	const className = cssClassNames(COMPONENT_NAME);

	const previousButtonMarkup = previousIcon && onPreviousClick && (
		<Button
			icon={previousIcon.icon}
			size="extra-small"
			onClick={onPreviousClick}
			clean
		/>
	);

	const nextButtonMarkup = nextIcon && onNextClick && (
		<Button
			icon={nextIcon.icon}
			size="extra-small"
			onClick={onNextClick}
			clean
		/>
	);

	const currentMonthLabel = MONTHS[month] ? (
		<div className={`${COMPONENT_NAME}__MonthLabel`}>
			<Text variant="label" element="span">
				{MONTHS[month]} {year}
			</Text>
		</div>
	) : null;

	return (
		<div className={className}>
			{previousButtonMarkup}
			{currentMonthLabel}
			{nextButtonMarkup}
		</div>
	);
};

export default Header;
