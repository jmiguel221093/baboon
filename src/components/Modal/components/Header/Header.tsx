import { cssClassNames, XIcon } from "../../../../utils";
import { Button } from "../../../Button";
import { Text } from "../../../Text";
import type { HeaderProps } from "./Header.props";

import "./Header.styles.scss";

const COMPONENT_NAME = "Modal__Header";

const Header = ({ title, onClose, closeIcon, metaData }: HeaderProps) => {
	const className = cssClassNames(COMPONENT_NAME);

	const dismissButtonMarkup = onClose ? (
		<div className={`${COMPONENT_NAME}__DismissButton`}>
			<Button
				icon={closeIcon || XIcon}
				size="small"
				clean
				onClick={onClose}
			/>
		</div>
	) : null;

	const titleActionsMarkup =
		title || metaData || dismissButtonMarkup ? (
			<div className={`${COMPONENT_NAME}__TitleActionsWrapper`}>
				<div
					// eslint-disable-next-line max-len
					className={`${COMPONENT_NAME}__TitleActionsWrapper__HeaderMainInfoWrapper`}
				>
					{metaData && (
						<div
							// eslint-disable-next-line max-len
							className={`${COMPONENT_NAME}__TitleActionsWrapper__HeaderMainInfoWrapper__MetaData`}
						>
							{typeof metaData === "string" ? (
								/* eslint-disable no-extra-parens */
								<Text variant="label" color="neutral">
									{metaData}
								</Text>
							) : (
								metaData
							)}
						</div>
						/* eslint-enable no-extra-parens */
					)}
					{title && <Text variant="displaySmall">{title}</Text>}
				</div>
				{dismissButtonMarkup}
			</div>
		) : null;

	return <div className={className}>{titleActionsMarkup}</div>;
};

export default Header;
