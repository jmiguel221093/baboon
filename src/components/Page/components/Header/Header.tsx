import { cssClassNames } from "../../../../utils";
import { Button } from "../../../Button";
import { Stack } from "../../../Stack";

import type { ActionType } from "../../../../types";

import { Title } from "../Title";

import type { HeaderProps } from "./Header.props";

import "./Header.styles.scss";
import { Thumbnail } from "../../../Thumbnail";

const COMPONENT_NAME = "Page__Header";

const Header = ({
	primaryAction,
	secondaryActions,
	breadcrumb,
	metadata,
	thumbnail,
	...rest
}: HeaderProps) => {
	const { subtitle, title } = rest;

	const className = cssClassNames(COMPONENT_NAME);

	const titleMarkup =
		title || subtitle ? (
			<div className={`${COMPONENT_NAME}__TitleWrapper`}>
				<Title subtitle={subtitle} title={title} />
			</div>
		) : null;

	const primaryActionMarkup = primaryAction ? (
		<Button
			primary
			onClick={primaryAction.onPress}
			disabled={primaryAction.disabled}
			loading={primaryAction.loading}
		>
			{primaryAction.content}
		</Button>
	) : null;

	const secondaryActionsMarkup =
		secondaryActions && secondaryActions.length > 0
			? secondaryActions.map(renderSecondaryActions)
			: null;

	const actionsMarkup =
		primaryActionMarkup || secondaryActionsMarkup ? (
			<div className={`${COMPONENT_NAME}__ActionsWrapper`}>
				<Stack alignment="center" distribution="start" spacing="medium">
					{secondaryActionsMarkup}
					{primaryActionMarkup}
				</Stack>
			</div>
		) : null;

	const actionsTitleWrapperMarkup = titleMarkup ? (
		<div className={`${COMPONENT_NAME}__ActionsTitleWrapper`}>
			{titleMarkup}
			{actionsMarkup}
		</div>
	) : null;

	const metadataMarkup = metadata ? (
		<div className={`${COMPONENT_NAME}__MetadataWrapper`}>{metadata}</div>
	) : null;

	const contentWrapperMarkup =
		actionsTitleWrapperMarkup || metadataMarkup ? (
			<div className={`${COMPONENT_NAME}__ContentWrapper`}>
				{actionsTitleWrapperMarkup}
				{metadataMarkup}
			</div>
		) : null;

	const thumbnailMarkup = thumbnail ? (
		<div className={`${COMPONENT_NAME}__ThumbnailWrapper`}>
			<Thumbnail {...thumbnail} />
		</div>
	) : null;

	const breadcrumbMarkup = breadcrumb ? (
		<div className={`${COMPONENT_NAME}__BreadcrumbWrapper`}>
			<Button
				icon={breadcrumb.icon}
				outline
				onClick={breadcrumb.onPress}
			/>
		</div>
	) : null;

	return (
		<div className={className}>
			{breadcrumbMarkup}
			{thumbnailMarkup}
			{contentWrapperMarkup}
		</div>
	);
};

function renderSecondaryActions(action: ActionType, index: number) {
	const { content, onPress: onClick, ...props } = action;
	return (
		<Button key={index} outline onClick={onClick} {...props}>
			{content}
		</Button>
	);
}

export default Header;
