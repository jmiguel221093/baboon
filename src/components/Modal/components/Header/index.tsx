import Header from "./Header";

export * from "./Header.props";
export { Header };
