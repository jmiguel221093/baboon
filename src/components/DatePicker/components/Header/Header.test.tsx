import { shallow } from "enzyme";
import Header from "./Header";
import type { DatePickerHeaderProps } from "./Header.props";

const LeftChevron = () => (
	<svg width="24" height="24" viewBox="0 0 24 24">
		<path
			fillRule="evenodd"
			clipRule="evenodd"
			d="M15.9999 21.5001C15.6025 21.4987 15.2216 21.3406 14.9399 21.0601L6.93993 13.0601C6.65656 12.7802 6.49707 12.3984 6.49707 12.0001C6.49707 11.6018 6.65656 11.22 6.93993 10.9401L14.9399 2.9401C15.3139 2.5388 15.877 2.37361 16.4085 2.50933C16.94 2.64506 17.355 3.06006 17.4907 3.59153C17.6264 4.12299 17.4612 4.68616 17.0599 5.0601L10.1199 12.0001L17.0599 18.9401C17.3433 19.22 17.5028 19.6018 17.5028 20.0001C17.5028 20.3984 17.3433 20.7802 17.0599 21.0601C16.7783 21.3406 16.3974 21.4987 15.9999 21.5001Z"
		/>
	</svg>
);

const RightChevron = () => (
	<svg width="24" height="24" viewBox="0 0 24 24">
		<path
			fillRule="evenodd"
			clipRule="evenodd"
			d="M6.21036 20.2104C6.21036 19.8062 6.36405 19.4019 6.67247 19.0935L13.9777 11.7893L6.67247 4.48508C6.05668 3.86824 6.05668 2.86824 6.67247 2.25139C7.28931 1.6356 8.28931 1.6356 8.90616 2.25139L17.3272 10.6725C17.943 11.2893 17.943 12.2893 17.3272 12.9061L8.90616 21.3272C8.28931 21.943 7.28931 21.943 6.67247 21.3272C6.36405 21.0188 6.21036 20.6146 6.21036 20.2104Z"
		/>
	</svg>
);

describe("Header component", () => {
	const defaultProps: DatePickerHeaderProps = {
		month: 4,
		year: 2022,
		onPreviousClick: jest.fn(),
		onNextClick: jest.fn(),
	};

	it("should render correctly with default props", () => {
		const wrapper = shallow(<Header {...defaultProps} />);
		expect(wrapper).toMatchSnapshot();
	});

	it("should render previous button when previousIcon and onPreviousClick props are provided", () => {
		const props = {
			...defaultProps,
			previousIcon: { icon: LeftChevron },
			onPreviousClick: jest.fn(),
		};
		const wrapper = shallow(<Header {...props} />);
		expect(wrapper.find("Button").length).toEqual(1);
		expect(wrapper.find("Button").prop("icon")).toEqual(LeftChevron);
	});

	it("should not render previous button when previousIcon prop is not provided", () => {
		const props = {
			...defaultProps,
			previousIcon: undefined,
		};
		const wrapper = shallow(<Header {...props} />);
		expect(wrapper.find("Button").length).toEqual(0);
	});

	it("should not render previous button when onPreviousClick prop is not provided", () => {
		const props = {
			...defaultProps,
			onPreviousClick: undefined,
		};
		const wrapper = shallow(<Header {...props} />);
		expect(wrapper.find("Button").length).toEqual(0);
	});

	it("should render next button when nextIcon and onNextClick props are provided", () => {
		const props = {
			...defaultProps,
			nextIcon: { icon: RightChevron },
			onNextClick: jest.fn(),
		};
		const wrapper = shallow(<Header {...props} />);
		expect(wrapper.find("Button").length).toEqual(1);
		expect(wrapper.find("Button").prop("icon")).toEqual(RightChevron);
	});

	it("should not render next button when nextIcon prop is not provided", () => {
		const props = {
			...defaultProps,
			nextIcon: undefined,
		};
		const wrapper = shallow(<Header {...props} />);
		expect(wrapper.find("Button").length).toEqual(0);
	});

	it("should not render next button when onNextClick prop is not provided", () => {
		const props = {
			...defaultProps,
			onNextClick: undefined,
		};
		const wrapper = shallow(<Header {...props} />);
		expect(wrapper.find("Button").length).toEqual(0);
	});
});
