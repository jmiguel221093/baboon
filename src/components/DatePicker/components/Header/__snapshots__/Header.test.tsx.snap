// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render correctly with default props 1`] = `ShallowWrapper {}`;
